dialogue-core/src/main/java/com/palantir/dialogue/core/StickySessioning.java

```diff
@@ -0,0 +1,120 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.collect.ListMultimap;
+import com.google.common.util.concurrent.FluentFuture;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.TagKey;
+import java.io.InputStream;
+import java.util.Optional;
+
+public final class StickySessioning {
+    private static final TagKey<LimitedChannel> EXECUTED_ON = new TagKey<>(LimitedChannel.class);
+
+    private StickySessioning() {}
+
+    static LimitedChannel wrapWithRecordedChoiceSelection(LimitedChannel delegate) {
+        return new RecordedChoiceChannel(delegate);
+    }
+
+    static ListenableFuture<Response> addExecutedOnTag(
+            ListenableFuture<Response> responseFuture,
+            LimitedChannel channel) {
+        return Futures.transform(responseFuture,
+                response -> new Response() {
+                    @Override
+                    public InputStream body() {
+                        return response.body();
+                    }
+
+                    @Override
+                    public int code() {
+                        return response.code();
+                    }
+
+                    @Override
+                    public ListMultimap<String, String> headers() {
+                        return response.headers();
+                    }
+
+                    @Override
+                    public void close() {
+                        response.close();
+                    }
+
+                    @Override
+                    public <T> Optional<T> getTag(TagKey<T> tagKey) {
+                        if (tagKey == EXECUTED_ON) {
+                            T ret = (T) channel;
+                            return Optional.of(ret);
+                        }
+                        return response.getTag(tagKey);
+                    }
+                }, MoreExecutors.directExecutor());
+    }
+
+    private static final class RecordedChoiceChannel implements LimitedChannel {
+        private final LimitedChannel delegate;
+
+        private RecordedChoiceChannel(LimitedChannel delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public Optional<ListenableFuture<Response>> maybeExecute(
+                Endpoint endpoint, Request request) {
+            return request.getTag(EXECUTED_ON).orElse(delegate).maybeExecute(endpoint, request);
+        }
+    }
+
+    private static final class StickySessionChannel implements Channel {
+        private final Channel delegate;
+
+        private FluentFuture<LimitedChannel> stuckChannel;
```

Comment:
I spent 10 minutes just perusing the codebase, and reading through the generated client code too, it seems to me like you give a channel to the generated client and it keeps it.

So I guess we'd have to create this channel at the start of a sticky session, and then from the sticky channel create a generated client. Then we'd get load balancing, but each session would be sticky.

That's how I understand this would be constructed.

Reply:
The flow is a bit mysterious for what this is trying to achieve, you let a request execute, then you remember that in the request at a special tag, then you remember that in here.